EIPS/eip-00.md

```diff
@@ -0,0 +1,790 @@
+        validAfter = (nonce >> 160) & 0xffffffffffff
+        assert Date.now() <= validUntil
+        assert Date.now() >= validAfter
+    }
+
+    if (sender.code.length == 0 && deployerData.length > 0) {
+        assert deployerData.length > 20
+        deployer := deployerData[0:20]
+        retDeployer, error := evm.Call(
+            from: AA_SENDER_CREATOR,
+            to: deployer,
+            input: deployerData[20:],
+            gas: validationGasLimit)
+        assert error == nil
+        assert sender.code.length > 0
+    }
+
+    if (paymasterData.length > 0) {
+        assert paymasterData.length > 20
+        paymaster := paymasterData[0:20]
+        paymasterInput := ABI.encodeWithSelector('validatePaymasterTransaction', tx, tx.hash)
+        retPaymaster, error := evm.Call(
+            from: AA_ENTRY_POINT,
+            to: paymaster,
+            input: paymasterInput,
+            gas: paymasterGasLimit)
+        assert error == nil
+        assert Date.now() <= retPaymaster.validUntil
+        assert Date.now() >= retPaymaster.validAfter
+        assert retPaymaster.isValid
+    }
+
+    if (sender.code.length == 0) {
+      signer := ecrecover(tx.hash, tx.signature)
+      assert signer == sender.address
+    } else {
+      senderInput := ABI.encodeWithSelector('validateTransaction', tx, tx.hash);
+      retSender, error := evm.Call(
+          from: AA_ENTRY_POINT,
+          to: sender,
+          input: senderInput,
+          gas: validationGasLimit - retDeployer.gasUsed)
+      assert error == nil
+      assert Date.now() <= retSender.validUntil
+      assert Date.now() >= retSender.validAfter
+      assert retSender.isValid
+    }
+}
+
+```
+
+In order to defend from DoS attack vectors, the block builders performing mempool transaction validation SHOULD consider
+the opcode banning and storage access rules described in EIP-9999.
+
+[Block validation](#execution-layer-block-validation) takes roughly the same amount of work as without AA transactions.
+In any case, validation must execute the entire block in order to verify the state change.
+During this execution, it currently verifies signatures, nonces, and gas payment.
+With Account Abstraction, it will also verify that all the validation frames were successful.
+There is a slight increase in required memory mostly used to store the `context` value that is passed from
+the `paymaster` validation frame to its post-transaction frame.
+
+As long as all transaction validation steps return correct values the block is considered valid.
+Block builders who are willing to relax the rules applied to the validation frames MAY do so.
+
+Such transactions MUST NOT be propagated through the default transaction mempool as they will be rejected by the nodes
+and the sending node will be blocked as a spammer.
+They may be propagated in the alternative mempool that allows them explicitly as defined in EIP-9999.
+
+### All validation state changes apply before all execution ones
```

Comment:
I don't fully understand the motivation behind this requirement

Each `AA_TX_TYPE` txn in the mempool gets simulated beforehand and each piece of state read during validation can be stored alongside the txn. During block building:

* Builder can gather `AA_TX_TYPE` txns, and (in parallel) check each of their associated validation state reads against the state after the last block, dropping those where state has changed (similar to conditional txns).
* Builder sequentially builds the block, keeping track of state changes. When including an `AA_TX_TYPE` txn the builder checks that the state reads haven't changed in the block. If yes, apply the txn (validation and execution in sequence)
* Validation rules and mempool reputation should protect the builder from too many drops due to changed state and failed validation frames.

Are there other building algorithms that this is attempting to accommodate?

Reply:
In step 1, it shouldn't drop them.  It needs to retest.  It's different from conditional txns because those are used by bundlers, that are aware of their tx getting dropped in this case. Users don't expect their transaction to drop, and would have no control over it. For example, the transaction uses a paymaster so validation accesses the paymaster's balance or storage. The transaction waits a couple of blocks in mempool while other transactions use the paymaster and change its state. The transaction shouldn't be dropped as it's still valid.

Due to the above (not being able to drop transactions based on changed state), step 2 becomes exponentially hard. The builder may keep running into transactions that were valid at the beginning of the block but invalidated by the last added transaction.

The builder may try to avoid the issue by deferring transactions that read slots in their validation, that has changed in an earlier transaction in the block.  But that introduces a censorship vector.  An attacker would find a slot that you access in validation, such as your ERC20 balance in a token you wish to pay with, and send you 1 token to write to that state, paying a slightly higher fee than your transaction.  Your transaction gets pushed off every block as long as the attacker keeps doing this, so the cost of censoring you is as low as sending one transaction per block.

To avoid censorship, the builder would have to resimulate validation of each transaction just before including it, which re-introduces the DoS vector.

Reply:
This makes sense, thanks for the explanation!